agent/wsclient/client.go

```diff
@@ -0,0 +1,270 @@
+type ClientServer interface {
+	AddRequestHandler(RequestHandler)
+	MakeRequest(input interface{}) error
+	Connect() error
+	Serve() error
+	io.Closer
+}
+
+//go:generate mockgen.sh github.com/aws/amazon-ecs-agent/agent/wsclient ClientServer mock/$GOFILE
+
+// ClientServerImpl wraps commonly used methods defined in ClientServer interface.
+type ClientServerImpl struct {
+	AcceptInvalidCert  bool
+	Conn               WebsocketConn
+	CredentialProvider credentials.AWSCredentialProvider
+	Region             string
+	// RequestHandlers is a map from message types to handler functions of the
+	// form:
+	//     "FooMessage": func(message *ecsacs.FooMessage)
+	RequestHandlers map[string]RequestHandler
+	// URL is the full url to the backend, including path, querystring, and so on.
+	URL string
+	ClientServer
+	ServiceError
+	TypeDecoder
+}
+
+// Connect opens a connection to the backend and upgrades it to a websocket. Calls to
+// 'MakeRequest' can be made after calling this, but responss will not be
+// receivable until 'Serve' is also called.
+func (cs *ClientServerImpl) Connect() error {
+	parsedURL, err := url.Parse(cs.URL)
+	if err != nil {
+		return err
+	}
+
+	signer := authv4.NewHttpSigner(cs.Region, ServiceName, cs.CredentialProvider, nil)
+
+	// NewRequest never returns an error if the url parses and we just verified
+	// it did above
+	request, _ := http.NewRequest("GET", cs.URL, nil)
+	signer.SignHttpRequest(request)
+
+	// url.Host might not have the port, but tls.Dial needs it
+	dialHost := parsedURL.Host
+	if !strings.Contains(dialHost, ":") {
+		dialHost += ":443"
+	}
+
+	timeoutDialer := &net.Dialer{Timeout: wsConnectTimeout}
+	log.Info("Creating poll dialer", "host", parsedURL.Host)
+	wsConn, err := tls.DialWithDialer(timeoutDialer, "tcp", dialHost, &tls.Config{InsecureSkipVerify: cs.AcceptInvalidCert})
+	if err != nil {
+		return err
+	}
+
+	websocketConn, httpResponse, err := websocket.NewClient(wsConn, parsedURL, request.Header, readBufSize, writeBufSize)
+	if httpResponse != nil {
+		defer httpResponse.Body.Close()
+	}
+	if err != nil {
+		var resp []byte
+		if httpResponse != nil {
+			var readErr error
+			resp, readErr = ioutil.ReadAll(httpResponse.Body)
+			if readErr != nil {
+				return errors.New("Unable to read websocket connection: " + readErr.Error() + ", " + err.Error())
+			}
+			// If there's a response, we can try to unmarshal it into one of the
+			// modeled error types
+			possibleError, _, decodeErr := DecodeData(resp, cs.TypeDecoder)
+			if decodeErr == nil {
+				return cs.NewError(possibleError)
+			}
+		}
+		log.Warn("Error creating a websocket client", "err", err)
+		return errors.New(string(resp) + ", " + err.Error())
+	}
+	cs.Conn = websocketConn
+	return nil
+}
+
+// AddRequestHandler adds a request handler to this client.
+// A request handler *must* be a function taking a single argument, and that
+// argument *must* be a pointer to a recognized 'ecsacs' struct.
```

Comment:
This comment is no longer accurate; just adding "recognized struct from decoder.GetRecognizedTypes(), e.g. 'ecsacs'" could fix it.